diff --git a/vowpalwabbit/options_boost_po.cc b/vowpalwabbit/options_boost_po.cc
index 100408d92..4abf2795a 100644
--- a/vowpalwabbit/options_boost_po.cc
+++ b/vowpalwabbit/options_boost_po.cc
@@ -163,25 +163,23 @@ std::string options_boost_po::help(const std::vector<std::string>& enabled_reduc
 {
   std::stringstream help;

-  // add general help
-  help << m_help_stringstream.find(m_default_tint)->second.rdbuf();
+  std::set<std::string> file_options = {"audit_regressor", "cache_file", "initial_regressor",
+      "input_feature_regularizer", "data", "output_feature_regularizer_text", "output_feature_regularizer_binary",
+      "final_regressor"};

-  // check if user only supplied --help or -h
-  if (m_supplied_options.size() <= 2)
+  for (const auto& kv : m_options)
   {
-    for (const auto& curr : m_help_stringstream)
+    help << "complete --command vw --long-option " << kv.second->m_name;
+    if (kv.second->m_short_name != "") { help << " --short-option " << kv.second->m_short_name; }
+    help << " --description '" << kv.first << ": " << kv.second->m_help << "'";
+    if (file_options.count(kv.second->m_name) > 0) { help << " --force-files "; }
+    else
     {
-      if (curr.first.compare(m_default_tint) != 0) help << curr.second.rdbuf();
-    }
-  }
-  else
-  {
-    // add help message of only enabled reductions
-    for (auto reduction : enabled_reductions)
-    {
-      auto it = m_help_stringstream.find(reduction);
-      if (it != m_help_stringstream.end()) { help << it->second.rdbuf(); }
+      help << " --no-files ";
     }
+
+    if (kv.second->m_type_hash != typeid(bool).hash_code()) { help << " --require-parameter "; }
+    help << "\n";
   }

   return help.str();
